src/repl/scala/tools/nsc/interpreter/ReplReporter.scala

```diff
@@ -50,7 +49,34 @@ class  ReplReporter(intp: IMain) extends ConsoleReporter(intp.settings, Console.
     printMessage(pos, prefix + msg)
   }
 
-  override def printMessage(msg: String) {
+  private lazy val indentDepth = replProps.promptText.lines.toList.last.length
+  private lazy val indentation = " " * indentDepth
+  private def indented(str: String) = str.lines.mkString(indentation, "\n" + indentation, "")
+
+  // indent errors, error message uses the caret to point at the line already on the screen instead of repeating it
+  // TODO: improve for multi-line -- can we splice the error into the code the user typed?
+  // (should also comment out the error to keep multi-line copy/pastable)
+  override def printMessage(posIn: Position, msg: String): Unit =
+   if ((posIn eq null) || !intp.printResults) super.printMessage(msg)
+   else {
+     val input = intp.currentRequest.originalLine
+     if (input.indexOf('\n') == -1) {
+       super.printMessage(indentation + posIn.lineCaret)
+       super.printMessage(indented(msg))
+     }
+     else {
+       val lineContent = posIn.lineContent
+
+       val preambleLines = intp.currentRequest.preambleEndPos.line
+
+       super.printMessage(s"On line ${posIn.line - preambleLines}:")
+       super.printMessage(indentation + lineContent)
+       super.printMessage(indentation + posIn.lineCaret)
+       super.printMessage(indented(msg))
+     }
+   }
```

Comment:
Great idea to lead with caret (and not with a stick). I'd forgotten this was part of that reversion. I see I needed to pick out pastes, which are not indented? For reference 05cc3e2271d2c2226ded33bef5a03f0c70c6f66d